python_modules/dagster/dagster/_core/definitions/sensor_definition.py

```diff
@@ -278,46 +487,55 @@ def _get_partitions_after_cursor(self, asset_key: AssetKey) -> List[str]:
             ]
         return partitions_to_fetch
 
+    def update_cursor_after_evaluation(self) -> None:
+        """Updates the cursor after the sensor evaluation function has been called. This method
+        should be called at most once per evaluation.
+        """
+
+        for fetched_event_id, advanced in self._event_fetched_is_advanced.items():
```

Comment:
Hmm, the current implementation of the cursor is to skip past unfetched events.

I also have a hard time thinking of why someone would want unfetched events to show up in their skipped events. It might also be annoying to have the "maximum skipped events" error thrown when those events were never requested.

Reply:
Trying to think of when an event might be unfetched:
- Calling `latest_materialization_records_by_key`, which will only fetch the latest event per asset. Future calls won't even pull up the skipped event because the cursor points to the latest evaluated event and any skipped event will be before that
- `materialization_records_for_key` is called sporadically and there are unfetched events in between each call. These events are presumably unpartitioned and since the user fetched updated events later, can't think of a reason why they would want old unfetched events
- User accidentally creates unpartitioned materialization for a partitioned asset. This event wouldn't show up in `latest_materialization_records_by_partition`/`latest_materialization_records_by_partition_and_asset`, but we could remedy this by also including `None` partitions in return value

Reply:
As general principle, I believe it's important to hard a strong separation between functions that mutate state that functions that retrieve it.  Without this separation, it makes it more difficult for users to reason about what's going on

> Hmm, the current implementation of the cursor is to skip past unfetched events.

Does that need to be the case? I believe that, from the user's perspective, we could make fetching and advancing independent.  Under the covers, `advance_cursor` (or `update_cursor_after_evaluation`) could fetch whatever data is needed in order to to do its thing, right? We could maintain a cache inside the context so that we don't fetch the same data twice?

Reply:
I can see a situation, for example if a user updates partitions sequentially and thus only updates partitions after the cursor via the `after_cursor_partition` flag, where storing unfetched materializations of prior partitions in the cursor is undesirable